Restart bugfix

nmsriram opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary
This merge fixes a parallel restart bug in the restart path.  Parallel restarts with same (or different) number of processors now work for the advection example.  

Also fixes a restart bug with reflective boundary conditions with an ugly hack in `mesh/mesh_refinement.cpp` to apply reflective boundary conditions.  It is expected that @Yurlungur or @jonahm-LANL will fix this when doing the boundary condition refactor.  As a result this should get merged in before the boundary condition fix.

We can now restart with a number of processors different from the original run, and even restart serial from parallel (or vice versa) and still get identical answers.

I believe #61 is fixed with this PR.

Known issue: 
<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->


## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md

JoshuaSBrown commented on src/mesh/mesh.cpp:
```diff
@@ -711,6 +708,9 @@ Mesh::Mesh(ParameterInput *pin, ApplicationInput *app_in, RestartReader &rr,
     bddisp = std::vector<int>(Globals::nranks);
   }
 
+  for (auto &c : costlist) {
+    c = 1.0;
+  }
```
Isn't this redundant with the initialization above?

pgrete replied:
I think so, too.

nmsriram replied:
This was me clutching at straws trying to figure out why MPI on more than original number of nodes still is incorrect and why the LANL physics code has the ghost cell problem.

nmsriram replied:
Removed.

JoshuaSBrown commented on src/mesh/mesh.hpp:
```diff
@@ -411,6 +411,7 @@ class Mesh {
       &UserWorkAfterLoopDefault;
   static void UserWorkInLoopDefault(); // called in main after each cycle
   std::function<void()> UserWorkInLoop = &UserWorkInLoopDefault;
+  int GetNbs() { return nslist[Globals::my_rank]; }
```
This should at least be made const.

nmsriram replied:
const wouldn't do anything since the method returns a scalar `int`, right?

Yurlungur replied:
I think @JoshuaSBrown means `int GetNBs() const { ... }`, which would enforce that the method does not change the state of the `Mesh` object.

JoshuaSBrown replied:
You are right, it would just show intent, but it will also allow you to call the method from within another const method. @Yurlungur thanks for clarifying.

nmsriram replied:
This has been removed.

JoshuaSBrown commented on src/mesh/mesh_refinement.hpp:
```diff
@@ -76,6 +76,16 @@ class MeshRefinement {
+  std::vector<int> GetInternals() {
+    return std::vector<int>(
+        {refine_flag_, neighbor_rflag_, deref_count_, deref_threshold_});
+  }
```
This can be made const noexcept.

nmsriram replied:
@JoshuaSBrown the restart test passes now in both serial and parallel, and I changed 'gold' to 'silver'.

nmsriram replied:
Also getInternals has been removed.

JoshuaSBrown commented on src/mesh/mesh_refinement.hpp:
```diff
@@ -76,6 +76,16 @@ class MeshRefinement {
+    return std::vector<int>(
+        {refine_flag_, neighbor_rflag_, deref_count_, deref_threshold_});
+  }
+  void SetInternals(std::vector<int> a) {
```
**a** should be const reference.

JoshuaSBrown commented on src/outputs/restart.hpp:
```diff
@@ -116,6 +116,7 @@ class RestartReader {
   std::vector<T> ReadDataset(const char *name, size_t *count = nullptr) {
     // Returns entire 1D array.
     // status, never checked.  We should...
+    std::vector<T> data;
```
Good catch! Oh nevermind, I guess this didn't actually really change anything...

pgrete replied:
Is this `ifdef` behavior actually what we want? (and/or is this a leftover from Athena++?)
I'm asking as having the code fail seems like a preferable solution rather than returning an empty object (due to lack of implementation or similar).

JoshuaSBrown commented on src/parthenon_manager.cpp:
```diff
@@ -197,7 +197,8 @@ void ParthenonManager::RestartPackages(Mesh &rm, RestartReader &resfile) {
   int nb = rm.GetNumMeshBlocksThisRank(Globals::my_rank);
-  int nbs = mb.gid;
+  int nbs = rm.GetNbs();
+  //  int nbs = mb.gid;
```
Remove comment?

nmsriram replied:
GetNbs() was removed and line uncommented.

JoshuaSBrown commented on src/mesh/mesh.cpp:
```diff
@@ -743,6 +743,17 @@ Mesh::Mesh(ParameterInput *pin, ApplicationInput *app_in, RestartReader &rr,
 
+  // Read in mesh refinement variables
+  if (adaptive) {
+    auto mrI = rr.ReadDataset<int32_t>("/Blocks/refinementInternals");
```
It looks like mrI can be made const.

nmsriram replied:
removed.

Yurlungur reviewed (approved):
Thanks for the fix. I don't love the function name `GetNbs` as `nb*` is a cryptic name. But since that's what they're called internally, I think it's fine.

JoshuaSBrown reviewed (commented):
I'm still getting errors when running the tests, which tests is this supposed to fix?

```Start 27: regression_test:advection_performance
27/36 Test #27: regression_test:advection_performance .............................***Failed  336.32 sec
      Start 28: regression_mpi_test:advection_performance
28/36 Test #28: regression_mpi_test:advection_performance .........................***Failed  412.27 sec
      Start 29: regression_test:restart
29/36 Test #29: regression_test:restart ...........................................   Passed   69.50 sec
      Start 30: regression_mpi_test:restart
30/36 Test #30: regression_mpi_test:restart .......................................***Failed    1.38 sec
```

JoshuaSBrown commented:
It looks like the test will run but something is still throwing an error:

```
/projects/opt/ppc64le/p9/openmpi/4.0.2-gcc_7.4.0/bin/mpiexec -n 2 '/home/joshuasbrown/Software/parthenon_x86_64/parthenon/build/example/advection/advection-example' -i '/home/joshuasbrown/Software/parthenon_x86_64/parthenon/tst/regression/test_suites/restart/parthinput.restart' 'parthenon/job/problem_id=gold'
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   cn2020
  Local device: mlx5_1
--------------------------------------------------------------------------
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
### Warning in Mesh::Initialize
The number of MeshBlocks increased more than twice during initialization.
More computing power than you expected may be required.
### Warning in Mesh::Initialize
The number of MeshBlocks increased more than twice during initialization.
More computing power than you expected may be required.

Setup complete, executing driver...

cycle=0 time=0.0000000000000000e+00 dt=1.7578125000000000e-03 zone-cycles/wsec = 0.00e+00
cycle=1 time=1.7578125000000000e-03 dt=1.7578125000000000e-03 zone-cycles/wsec = 1.85e+04
cycle=2 time=3.5156250000000001e-03 dt=1.7578125000000000e-03 zone-cycles/wsec = 1.36e+05
cycle=3 time=5.2734375000000003e-03 dt=1.7578125000000000e-03 zone-cycles/wsec = 6.90e+04
[cn2020:30009] 1 more process has sent help message help-mpi-btl-openib.txt / error in device init
[cn2020:30009] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

JoshuaSBrown reviewed (approved):
Other than my optional improvements this is good. The error I was running into is located in the python regression testing framework so this has my green light.

Yurlungur commented:
@nmsriram is this ready for merge?

pgrete commented on src/mesh/mesh_refinement.hpp:
```diff
@@ -76,6 +76,16 @@ class MeshRefinement {
   // and/or in BoundaryValues::ProlongateBoundaries() (for SMR and AMR)
   int AddToRefinement(ParArrayND<Real> pvar_cc, ParArrayND<Real> pcoarse_cc);
   int AddToRefinement(FaceField *pvar_fc, FaceField *pcoarse_fc);
+  std::vector<int> GetInternals() {
```
Do we need a `vector` here? Could this just be an `array`?

nmsriram replied:
This has been removed.

pgrete commented on src/mesh/mesh_refinement.hpp:
```diff
@@ -76,6 +76,16 @@ class MeshRefinement {
   // and/or in BoundaryValues::ProlongateBoundaries() (for SMR and AMR)
   int AddToRefinement(ParArrayND<Real> pvar_cc, ParArrayND<Real> pcoarse_cc);
   int AddToRefinement(FaceField *pvar_fc, FaceField *pcoarse_fc);
+  std::vector<int> GetInternals() {
```
Given that `int32_t` is used above to read, I suggest to also be more specific here.

nmsriram replied:
Will do.

nmsriram commented:
@pgrete the reason to return an empty array is that we may change the physics on restart and I didn't want the request for those physics to kill the code.  I could go either way on this, perhaps with a query function.

pgrete commented:
@nmsriram there are a couple of comments left. It'd be great if they could be addressed (or replied to) before we go ahead and merge this.

Also (unrelated to this PR): I just noticed that quite a bit of the restart functionality has been implemented using macros.
What's the motivation behind those macros and could we potentially replace those (not in this PR) with standard functions?

sriram-LANL committed:
added a hack for applying reflective boundary conditions after prolong / restrict

sriram-LANL committed:
added initializationf or sparse variable start

sriram-LANL committed:
Fixed mpi bug in restart.hpp where every rank was reading the first
N blocks instead of reading at offset for that particular rank.

Added capability to restart with or without ghost zones.

sriram-LANL committed:
Merge branch 'develop' into sriram/tmp2

sriram-LANL committed:
read in nbdel and nbnew from restart file.  Added deep copy to ensure GPU data is initialized on restart.

sriram-LANL committed:
parallel regression test now checks silver file instead of comparing gold to itself

sriram-LANL committed:
fixed formatting to satisfy cpplint and clang-format

sriram-LANL committed:
changed syntax for pointer from (*v). to v->

sriram-LANL committed:
Merge branch 'bugfix/restart2' of github.com:lanl/parthenon into bugfix/restart2

nmsriram commented:
@pgrete the reason they are macros is because I couldn't figure out how to do it different. About a week ago @AndrewGaspar commented essentially the same thing and showed a std::function way.  I'm a bit hesitant to change this code at this point and will do it in a subsequent PR.

sriram-LANL committed:
updated changelog

sriram-LANL committed:
fixed PR number in changelog

sriram-LANL committed:
fixed make format errors

Yurlungur commented on src/interface/variable_pack.hpp:
```diff
@@ -154,7 +154,7 @@ void FillVarView(const vpack_types::VarList<T> &vars, PackIndexMap *vmap,
 
   int vindex = 0;
-  int sparse_start;
+  int sparse_start = -22000; // Initialize to an obviously wrong number
```
We should perhaps formalize this throughout the code and make a habit of initializing to [signal NaNs](https://en.cppreference.com/w/cpp/types/numeric_limits/signaling_NaN). That's for another time, though.

JoshuaSBrown replied:
Should we create an issue?

Yurlungur replied:
yes good idea

Yurlungur commented on src/mesh/mesh_refinement.cpp:
```diff
@@ -38,6 +38,52 @@
 
 namespace parthenon {
 
+// TODO(Yurlunger) Identify an idiomatic way of applying reflective
```
```suggestion
// TODO(JMM) Identify an idiomatic way of applying reflective
```

Yurlungur commented on src/mesh/mesh_refinement.cpp:
```diff
@@ -38,6 +38,52 @@
+  if (pmb->boundary_flag[BoundaryFace::outer_x1] == BoundaryFlag::reflect) {
+    for (int n = 0; n < a.GetDim(4); n++) {
+      for (int j = 0; j <= jb.e + NGHOST; j++) {
+        for (int i = ib.e + 1; i <= ib.e + NGHOST; i++) {
+          a(n, 0, j, i) = a(n, 0, j, 2 * ib.e - i + 1);
+        }
+      }
+    }
+  }
+
+  if (pmb->boundary_flag[BoundaryFace::inner_x1] == BoundaryFlag::reflect) {
+    for (int n = 0; n < a.GetDim(4); n++) {
+      for (int j = 0; j <= jb.e + NGHOST; j++) {
+        for (int i = 0; i < ib.s; i++) {
+          a(n, 0, j, i) = a(n, 0, j, 2 * ib.s - i - 1);
+        }
+      }
+    }
+  }
+
+  if (pmb->boundary_flag[BoundaryFace::outer_x2] == BoundaryFlag::reflect) {
+    for (int n = 0; n < a.GetDim(4); n++) {
+      for (int j = jb.e + 1; j <= jb.e + NGHOST; j++) {
+        for (int i = 0; i <= ib.e + NGHOST; i++) {
+          a(n, 0, j, i) = a(n, 0, 2 * jb.e - j + 1, i);
+        }
+      }
+    }
+  }
+
+  if (pmb->boundary_flag[BoundaryFace::inner_x2] == BoundaryFlag::reflect) {
+    for (int n = 0; n < a.GetDim(4); n++) {
+      for (int j = 0; j < jb.s; j++) {
+        for (int i = 0; i <= ib.e + NGHOST; i++) {
+          a(n, 0, j, i) = a(n, 0, 2 * jb.s - j - 1, i);
+        }
+      }
+    }
+  }
+}
```
This actually needs to be done for reflecting, outflow, and "user," not just reflecting. But I'll deal with it in my MR.

nmsriram replied:
Thanks!

Yurlungur commented on tst/regression/test_suites/restart/restart.py:
```diff
@@ -53,7 +53,7 @@ def Analyse(self, parameters):
         varName = "/advected"
         goldData = gold[varName][:].flatten()
-        silverData = gold[varName][:].flatten()
+        silverData = silver[varName][:].flatten()
```
This is a good and important fix. But it may conflict with the same fix in #330

nmsriram replied:
#330 makes a few more deeper changes to the structure and comparisons, and will override this one.  Without this fix in place I couldn't test which is why I made it here as well.

Yurlungur commented on src/parthenon_manager.cpp:
```diff
@@ -222,15 +230,13 @@ void ParthenonManager::RestartPackages(Mesh &rm, RestartReader &resfile) {
     const std::string vName = v->label();
 
-    std::cout << "Var:" << vName << ":" << v4 << std::endl;
+    if (Globals::my_rank == 0) std::cout << "Var:" << vName << ":" << v4 << std::endl;
```
is this a debug statement that should be removed?

nmsriram replied:
I put it in there intentionally.  I wanted to give out a status reading that variables are read in from restart file so that the outputs from a direct run and a restart run are evident, and also so that somebody can check to see what variables came in from the restart file.

Yurlungur replied:
Ah, okay. Got it.

Yurlungur reviewed (approved):
nice work. this one was a doozy.

pgrete commented on src/parthenon_manager.cpp:
```diff
@@ -192,7 +192,9 @@ ParthenonManager::ProcessPackagesDefault(std::unique_ptr<ParameterInput> &pin) {
   // Restart packages with information for blocks in ids from the restart file
   // Assumption: blocks are contiguous in restart file, may have to revisit this.
-  const IndexDomain interior = IndexDomain::interior;
+  //  const IndexDomain interior = IndexDomain::interior;
```
```suggestion
```

JoshuaSBrown commented on src/mesh/mesh_refinement.cpp:
```diff
@@ -129,6 +175,7 @@ void MeshRefinement::RestrictCellCenteredValues(const ParArrayND<Real> &fine,
                (fine(n, 0, j, i + 1) * vol01 + fine(n, 0, j + 1, i + 1) * vol11)) /
               tvol;
         });
+    applyBounds(pmb, coarse, cib, cjb);
```
Why do we only have a function like this for the 2d case?

Yurlungur replied:
good catch. I think this should be for all 3 cases, @nmsriram

nmsriram replied:
Oh brother. yes.  of course.  @JoshuaSBrown / @Yurlungur  Should I put in another PR to fix this, or should I just depend on @Yurlungur to fix it?

JoshuaSBrown replied:
Whoever, gets to it first I suppose.

JoshuaSBrown commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -198,33 +198,22 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
 
-  const IndexDomain interior = IndexDomain::interior;
-  const IndexDomain entire = IndexDomain::entire;
+  const IndexDomain theDomain =
+      (output_params.include_ghost_zones ? IndexDomain::entire : IndexDomain::interior);
```
Nice!

JoshuaSBrown commented on src/outputs/restart.cpp:
```diff
@@ -236,13 +240,14 @@ void RestartOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm)
-  auto &nx2 = mb.block_size.nx2;
-  auto &nx3 = mb.block_size.nx3;
-  int bsize[3] = {nx1, nx2, nx3};
+  auto nx1 = out_ib.e - out_ib.s + 1; // SS mb.block_size.nx1;
+  auto nx2 = out_jb.e - out_jb.s + 1; // SS mb.block_size.nx2;
+  auto nx3 = out_kb.e - out_kb.s + 1; // SS mb.block_size.nx3;
+  int bsize[3] = {mb.block_size.nx1, mb.block_size.nx2, mb.block_size.nx3};
```
These comments are confusing to me, what does the SS mean, this kind of looks like you are saying nx1 is the same as mb.block_size.nx1. If so why are you using mb.block_size.nx1 to define bsize and not the more concise nx1 nx2 and nx3 variables? 

Ok, I see, nx1 etc will change depending on theDomain.

nmsriram replied:
Sorry, SS are my initials and I use // SS or //SS where I think I may have screwed up and may need to go back.

JoshuaSBrown commented on src/outputs/restart.hpp:
```diff
@@ -180,6 +179,9 @@ class RestartReader {
   void Close();
 
+  // Does file have ghost cells?
+  int hasGhost;
```
Why int instead of bool?

nmsriram replied:
Because it is read in from the HDF file and this requires fewer lines of code

JoshuaSBrown commented:
```
    Start  1: unit_tests:Just check everything
 1/37 Test  #1: unit_tests:Just check everything ..................................   Passed    0.27 sec
      Start  2: unit_tests:Task Object Lifecycle
 2/37 Test  #2: unit_tests:Task Object Lifecycle ..................................   Passed    0.27 sec
      Start  3: unit_tests:Can create a vector-valued face-variable
 3/37 Test  #3: unit_tests:Can create a vector-valued face-variable ...............   Passed    0.25 sec
      Start  4: unit_tests:Add and Get is called
 4/37 Test  #4: unit_tests:Add and Get is called ..................................   Passed    0.26 sec
      Start  5: unit_tests:reset is called
 5/37 Test  #5: unit_tests:reset is called ........................................   Passed    0.26 sec
      Start  6: unit_tests:when hasKey is called
 6/37 Test  #6: unit_tests:when hasKey is called ..................................   Passed    0.29 sec
      Start  7: unit_tests:Physical constants
 7/37 Test  #7: unit_tests:Physical constants .....................................   Passed    0.29 sec
      Start  8: unit_tests:Checking IndexShape indices
 8/37 Test  #8: unit_tests:Checking IndexShape indices ............................   Passed    0.25 sec
      Start  9: unit_tests:Checking IndexShape cell counts
 9/37 Test  #9: unit_tests:Checking IndexShape cell counts ........................   Passed    0.24 sec
      Start 10: unit_tests:par_for loops
10/37 Test #10: unit_tests:par_for loops ..........................................   Passed    0.61 sec
      Start 11: unit_tests:nested par_for loops
11/37 Test #11: unit_tests:nested par_for loops ...................................   Passed    0.29 sec
      Start 12: unit_tests:Overlapping SpaceInstances
12/37 Test #12: unit_tests:Overlapping SpaceInstances .............................   Passed    0.34 sec
      Start 13: unit_tests:Built-in flags are registered
13/37 Test #13: unit_tests:Built-in flags are registered ..........................   Passed    0.34 sec
      Start 14: unit_tests:A Metadata flag is allocated
14/37 Test #14: unit_tests:A Metadata flag is allocated ...........................   Passed    0.26 sec
      Start 15: unit_tests:A Metadata struct is created
15/37 Test #15: unit_tests:A Metadata struct is created ...........................   Passed    0.25 sec
      Start 16: unit_tests:ParArrayND
16/37 Test #16: unit_tests:ParArrayND .............................................   Passed    0.31 sec
      Start 17: unit_tests:ParArrayND with LayoutLeft
17/37 Test #17: unit_tests:ParArrayND with LayoutLeft .............................   Passed    0.35 sec
      Start 18: unit_tests:Time simple stencil operations
18/37 Test #18: unit_tests:Time simple stencil operations .........................   Passed    0.27 sec
      Start 19: unit_tests:Check registry pressure
19/37 Test #19: unit_tests:Check registry pressure ................................   Passed    0.99 sec
      Start 20: unit_tests:Check many arrays
20/37 Test #20: unit_tests:Check many arrays ......................................   Passed    0.26 sec
      Start 21: unit_tests:Swarm memory management
21/37 Test #21: unit_tests:Swarm memory management ................................   Passed    0.31 sec
      Start 22: unit_tests:Can pull variables from containers based on Metadata
22/37 Test #22: unit_tests:Can pull variables from containers based on Metadata ...   Passed    0.67 sec
      Start 23: unit_tests:Test required/desired checking from inputs
23/37 Test #23: unit_tests:Test required/desired checking from inputs .............   Passed    0.26 sec
      Start 24: unit_tests:Parthenon Error Checking
24/37 Test #24: unit_tests:Parthenon Error Checking ...............................   Passed    0.27 sec
      Start 25: unit_tests:Check that integer division, rounded up, works
25/37 Test #25: unit_tests:Check that integer division, rounded up, works .........   Passed    0.27 sec
      Start 26: unit_tests:Check that partitioning a container works
26/37 Test #26: unit_tests:Check that partitioning a container works ..............   Passed    0.26 sec
      Start 27: performance_tests:Catch2 Container Iterator Performance
27/37 Test #27: performance_tests:Catch2 Container Iterator Performance ...........   Passed   11.80 sec
      Start 28: regression_test:advection_performance
28/37 Test #28: regression_test:advection_performance .............................   Passed  334.83 sec
      Start 29: regression_mpi_test:advection_performance

29/37 Test #29: regression_mpi_test:advection_performance .........................   Passed  1095.79 sec
      Start 30: regression_test:restart
30/37 Test #30: regression_test:restart ...........................................   Passed   69.44 sec
      Start 31: regression_mpi_test:restart
31/37 Test #31: regression_mpi_test:restart .......................................   Passed  112.76 sec
      Start 32: regression_test:calculate_pi
32/37 Test #32: regression_test:calculate_pi ......................................   Passed    2.77 sec
      Start 33: regression_mpi_test:calculate_pi
33/37 Test #33: regression_mpi_test:calculate_pi ..................................   Passed    2.92 sec
      Start 34: regression_test:advection_convergence
34/37 Test #34: regression_test:advection_convergence .............................   Passed  376.76 sec
      Start 35: regression_mpi_test:advection_convergence
35/37 Test #35: regression_mpi_test:advection_convergence .........................   Passed  682.34 sec
      Start 36: regression_test:output_hdf5
36/37 Test #36: regression_test:output_hdf5 .......................................   Passed  241.76 sec
      Start 37: regression_mpi_test:output_hdf5
37/37 Test #37: regression_mpi_test:output_hdf5 ...................................   Passed  368.85 sec

```
Congrats! It works.

Yurlungur reviewed (approved):
not sure why my approval didn't stick. I approve.
